src/System.Text.Json/src/System/Text/Json/JsonDocument.Parse.cs

```diff
@@ -0,0 +1,280 @@
+
+            if (utf8Json.IsSingleSegment)
+            {
+                return Parse(utf8Json.First, readerOptions, null);
+            }
+
+            int length = checked((int)utf8Json.Length);
+            byte[] utf8Bytes = ArrayPool<byte>.Shared.Rent(length);
+
+            try
+            {
+                utf8Json.CopyTo(utf8Bytes.AsSpan());
+                return Parse(utf8Bytes.AsMemory(0, length), readerOptions, utf8Bytes);
+            }
+            catch
+            {
+                // Holds document content, clear it before returning it.
+                utf8Bytes.AsSpan(0, length).Clear();
+                ArrayPool<byte>.Shared.Return(utf8Bytes);
+                throw;
+            }
+        }
+
+        public static JsonDocument Parse(Stream utf8Json, JsonReaderOptions readerOptions = default)
+        {
+            if (utf8Json == null)
+                throw new ArgumentNullException(nameof(utf8Json));
+
+            CheckSupportedOptions(readerOptions);
+
+            ArraySegment<byte> drained = ReadToEnd(utf8Json);
+
+            try
+            {
+                return Parse(drained.AsMemory(), readerOptions, drained.Array);
+            }
+            catch
+            {
+                // Holds document content, clear it before returning it.
+                drained.AsSpan().Clear();
+                ArrayPool<byte>.Shared.Return(drained.Array);
+                throw;
+            }
+        }
+
+        public static Task<JsonDocument> ParseAsync(
+            Stream utf8Json,
+            JsonReaderOptions readerOptions = default,
+            CancellationToken cancellationToken = default)
+        {
+            if (utf8Json == null)
+                throw new ArgumentNullException(nameof(utf8Json));
+
+            CheckSupportedOptions(readerOptions);
+
+            return ParseAsyncCore(utf8Json, readerOptions, cancellationToken);
+        }
+
+        private static async Task<JsonDocument> ParseAsyncCore(
+            Stream utf8Json,
+            JsonReaderOptions readerOptions = default,
+            CancellationToken cancellationToken = default)
+        {
+            ArraySegment<byte> drained = await ReadToEndAsync(utf8Json, cancellationToken).ConfigureAwait(false);
+
+            try
+            {
+                return Parse(drained.AsMemory(), readerOptions, drained.Array);
+            }
+            catch
+            {
+                // Holds document content, clear it before returning it.
+                drained.AsSpan().Clear();
+                ArrayPool<byte>.Shared.Return(drained.Array);
+                throw;
+            }
+        }
+
+        public static JsonDocument Parse(ReadOnlyMemory<char> json, JsonReaderOptions readerOptions = default)
+        {
+            CheckSupportedOptions(readerOptions);
+
+            ReadOnlySpan<char> jsonChars = json.Span;
+            int byteCount = Utf8JsonReader.Utf8Encoding.GetByteCount(jsonChars);
```

Comment:
Rather than first getting the byte count and then transcoding, why not pessimistically rent a buffer of max size (i.e. 3x json.Length)?

Reply:
The maximum is 4x json.Length, not 3x.  Renting arrays that are significantly larger than the needed size can cause a temporary starvation for that bucket and cause other parts of the system to have to deal with an allocation/compaction (which is, of course, worse at larger sizes).  Counting seems fast enough, and callers could always choose to do the overalloc-transcode themselves.